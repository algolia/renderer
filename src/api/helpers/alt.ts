import Altheia from 'altheia-async-data-validator';

export const alt = Altheia.instance();
alt.lang('protocol_not_allowed', () => 'Only HTTP protocol is allowed');

export function getDefaultParams(): Record<any, any> {
  return {
    url: alt
      .internet()
      .url()
      .custom('protocol_not_allowed', (val) => {
        return ['http:', 'https:'].includes(new URL(val).protocol);
      })
      .required(),
    ua: alt.string().required(),
    waitTime: alt.object().schema(
      alt({
        min: alt.number().cast().min(1000).max(19000),
        max: alt.number().cast().min(2000).max(20000),
      })
    ),
  };
}
