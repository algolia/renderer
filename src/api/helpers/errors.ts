import type express from 'express';

interface AnyParams {
  res: express.Response;
  status: number;
  message: string;
  details?: any;
}

function any({ res, status, message, details }: AnyParams): void {
  res.status(status).json({ error: true, message, details });
}

interface BadRequestParams {
  res: express.Response;
  message?: string;
  details?: any;
}

export function badRequest({
  res,
  details,
  message = 'Bad Request',
}: BadRequestParams): void {
  return any({
    res,
    status: 400,
    message,
    details,
  });
}
