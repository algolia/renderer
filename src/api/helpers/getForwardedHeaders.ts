import type express from 'express';

import { HEADERS_TO_FORWARD } from 'api/constants';

export function getForwardedHeadersFromRequest(
  req: express.Request<any, any, any, any>
): Record<string, string> {
  const headersToForward = HEADERS_TO_FORWARD.reduce((partial, headerName) => {
    const name = headerName.toLowerCase();
    if (req.headers[name]) {
      return { ...partial, [name]: req.headers[name] };
    }
    return partial;
  }, {});

  return headersToForward;
}
