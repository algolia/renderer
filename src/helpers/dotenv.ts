export default () => {
  if (process.env.NODE_ENV === 'development') {
    require('dotenv-safe').config({ allowEmptyValues: true });
  }
};
