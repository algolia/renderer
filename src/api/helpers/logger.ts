import { log as mainLog } from 'helpers/logger';

export const log = mainLog.child({ svc: 'api ' });
