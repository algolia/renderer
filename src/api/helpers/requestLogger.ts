import type express from 'express';

import { log } from './logger';

export function requestLogger(
  req: express.Request,
  res: express.Response,
  next: express.NextFunction
): void {
  if (['/ready', '/healthy'].includes(req.url)) {
    next();
    return;
  }

  log.info('New request', { method: req.method, path: req.url });
  next();
}
