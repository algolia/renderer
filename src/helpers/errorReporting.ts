import * as Sentry from '@sentry/node';

import { log } from './logger';

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: process.env.npm_package_version,
  environment: process.env.CLUSTER_NAME || process.env.NODE_ENV,
  serverName: 'renderscript',
  ignoreErrors: [],
  maxBreadcrumbs: 10,
});

export function report(err: Error, extra: any = {}): void {
  if (!process.env.SENTRY_DSN) {
    log.error(err.message, err.stack, extra);
    return;
  }

  log.error(err.message);
  Sentry.withScope((scope) => {
    scope.setExtras(extra);
    Sentry.captureException(err);
  });
}

export async function drain(): Promise<boolean> {
  const client = Sentry.getCurrentHub().getClient();
  if (client) {
    return await client.close(2000);
  }

  return true;
}
