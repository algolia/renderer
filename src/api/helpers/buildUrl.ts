import { report } from 'helpers/errorReporting';

const DOCKER_LOCALHOST = 'host.docker.internal';

const USE_DOCKER_LOCALHOST = process.env.USE_DOCKER_LOCALHOST === 'true';

export function replaceHost(url: URL, from: string, to: string): URL {
  const fromRegex = new RegExp(`^${from}(:|$)`);
  const host = url.host || '';
  // eslint-disable-next-line no-param-reassign
  url.host = host.replace(fromRegex, `${to}$1`);
  return url;
}

export function revertUrl(href: string | null): URL | null {
  if (!href) {
    return null;
  }

  try {
    const url = new URL(href);
    if (!USE_DOCKER_LOCALHOST) {
      return url;
    }
    return replaceHost(url, DOCKER_LOCALHOST, 'localhost');
  } catch (err) {
    report(new Error('invalid revertUrl'), { href });
    return null;
  }
}

export function buildUrl(href: string): URL {
  const url = new URL(href);
  if (!USE_DOCKER_LOCALHOST) {
    return url;
  }
  return replaceHost(url, 'localhost', DOCKER_LOCALHOST);
}
