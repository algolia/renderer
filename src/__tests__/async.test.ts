import type { PostRenderSuccess } from 'api/@types/postRender';

import { cleanString, postRender, request } from './helpers';

jest.setTimeout(10000);

describe('async', () => {
  it.each(['chromium', 'firefox'])(
    'should render async page on %s',
    async (browser) => {
      const { res, body } = await request(
        `http://localhost:3000/render?url=http%3A%2F%2Flocalhost%3A3000%2Ftest-website%2Fasync.html&ua=Algolia+Crawler&browser=${browser}`
      );

      expect(res.statusCode).toBe(200);
      expect(res.headers).toEqual({
        connection: 'keep-alive',
        'content-length': expect.any(String),
        'content-security-policy':
          "default-src 'none'; style-src * 'unsafe-inline'; img-src * data:; font-src *",
        'content-type': 'text/html; charset=utf-8',
        date: expect.any(String),
        etag: expect.any(String),
        'keep-alive': 'timeout=5',
      });

      expect(cleanString(body)).toMatchSnapshot();
    }
  );

  it('should wait by default for 0ms', async () => {
    const { res, body } = await postRender({
      url: 'http://localhost:3000/test-website/async.html',
      ua: 'Algolia Crawler',
    });

    const json: PostRenderSuccess = JSON.parse(body);
    expect(res.statusCode).toBe(200);
    expect(json.metrics.timings.total).toBeLessThanOrEqual(2000);
    expect(json.body).not.toMatch('4. setTimeout 1000');
  });

  it('should wait at least 6000ms', async () => {
    const { res, body } = await postRender({
      url: 'http://localhost:3000/test-website/async.html',
      ua: 'Algolia Crawler',
      waitTime: {
        min: 6000,
      },
    });

    const json: PostRenderSuccess = JSON.parse(body);
    expect(res.statusCode).toBe(200);

    expect(json.metrics.timings.minWait).toBeGreaterThanOrEqual(5000);
    expect(json.metrics.timings.total).toBeGreaterThanOrEqual(6000);
    expect(json.metrics.timings.total).toBeLessThanOrEqual(7000);
    expect(json.body).toMatch('5. setTimeout 5000');
  });

  it('should wait at most 5000ms', async () => {
    const { res, body } = await postRender({
      url: 'http://localhost:3000/test-website/slow.html',
      ua: 'Algolia Crawler',
      waitTime: {
        min: 4000,
        max: 5000,
      },
    });

    const json: PostRenderSuccess = JSON.parse(body);
    expect(res.statusCode).toBe(200);
    expect(json.metrics.timings.goto).toBeLessThanOrEqual(50);

    // In that case the page is slow so min is not used
    expect(json.metrics.timings.minWait).toBeNull();

    expect(json.metrics.timings.ready).toBeLessThanOrEqual(5020);
    expect(json.metrics.timings.total).toBeGreaterThanOrEqual(4000);
    expect(json.metrics.timings.total).toBeLessThanOrEqual(5120);

    // We count the dot because there is no way to have precise execution
    // There should be around 25 dots (one fetch every 200ms during 5s = 25 dots)
    // We check for 20 to have some margin
    // And no more than 30 to check that it was not executed more than 5s
    expect(json.body).toMatch('.'.repeat(20));
    expect(json.body).not.toMatch('.'.repeat(30));
  });
});
