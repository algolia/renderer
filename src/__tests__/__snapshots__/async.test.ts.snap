// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async should render async page on chromium 1`] = `"<!DOCTYPE html><html><head></head><body><div id="ua">Algolia Crawler</div><script>document.getElementById('ua').innerText = window.navigator.userAgent;</script><script>let i = 1;const addEvent = name => {document.body.innerHTML += \`\${i === 1 ? i : \` - \${i}\`}. \${name}\`;i++;}addEvent('Init')window.onload = addEvent.bind(null, 'window.onload');window.addEventListener('DOMContentLoaded', addEvent.bind(null, 'DOMContentLoaded'));setTimeout(addEvent.bind(null, 'setTimeout 2000'), 2000);setTimeout(addEvent.bind(null, 'setTimeout 5000'), 5000);setTimeout(addEvent.bind(null, 'setTimeout 10000'), 10000);setTimeout(addEvent.bind(null, 'setTimeout 20000'), 20000);</script>1. Init - 2. DOMContentLoaded - 3. window.onload</body></html>"`;

exports[`async should render async page on firefox 1`] = `"<!DOCTYPE html><html><head></head><body><div id="ua">Algolia Crawler</div><script>document.getElementById('ua').innerText = window.navigator.userAgent;</script><script>let i = 1;const addEvent = name => {document.body.innerHTML += \`\${i === 1 ? i : \` - \${i}\`}. \${name}\`;i++;}addEvent('Init')window.onload = addEvent.bind(null, 'window.onload');window.addEventListener('DOMContentLoaded', addEvent.bind(null, 'DOMContentLoaded'));setTimeout(addEvent.bind(null, 'setTimeout 2000'), 2000);setTimeout(addEvent.bind(null, 'setTimeout 5000'), 5000);setTimeout(addEvent.bind(null, 'setTimeout 10000'), 10000);setTimeout(addEvent.bind(null, 'setTimeout 20000'), 20000);</script>1. Init - 2. DOMContentLoaded - 3. window.onload</body></html>"`;
